### PR TITLE
snap: show nicer message when we get "access denied" from the server

### PR DESCRIPTION
**Pull request by mvo5**
Strawman branch for a nicer `access denied` message.

Closes: LP#1574829

### CODE REVIEW

File: cmd/snap/main.go [hunk 1]
```diff
@@ -129,11 +129,19 @@ func init() {
 	}
 }
 
+var accessDeniedHelp = `Access denied.
+
+Please use "sudo" or the "snap login <userid>" command
```
Comment (chipaca) [inline cmd/snap/main.go]: At the risk of bikeshedding, can we make it more evident that we recommend doing `snap login`? I'd go as far as not even mentioning sudo, but maybe that's just me. With sudo some things won't work properly *anyway*, as what's really needed is to be logged in.

**Comment by chipaca** [comment]
:+1:, mostly.

File: cmd/snap/main.go (hunk 1 above)
Comment (mvo5) [inline cmd/snap/main.go]: Thats fine with me, I'm happy to update the string. Maybe:
```Access denied.
Please use "snap login <your ubuntu sso account>".
```
(is there something nicer than `Access denied.` in the first line?

Comment (niemeyer) [inline cmd/snap/main.go]: Can I join the bikeshed? :-)

`Access denied (snap login --help).`

Then, let's fix snap login to be more clear:

```
type cmdLogin struct {
        Positional struct {
                // FIXME: add support for translated descriptions
                //        (see cmd/snappy/common.go:addOptionDescription)
                UserName string `positional-arg-name:"email" description:"login.ubuntu.com email to login as"`
        } `positional-args:"yes" required:"yes"`
}

var shortLoginHelp = i18n.G("Authenticates on snapd and the store")

var longLoginHelp = i18n.G(`
The login command authenticates on snapd and the snap store and saves credentials
into the ~/.snap/auth.json file. Further communication with snapd will then be made
using those credentials.

Accounts may be created at https://login.ubuntu.com.
`)
```

Comment (niemeyer) [inline cmd/snap/main.go]: (made a couple of minor updates above)

Comment (niemeyer) [inline cmd/snap/main.go]: (more updates)

File: cmd/snap/main.go [hunk 2]
```diff
@@ -129,11 +129,19 @@ func init() {
...
+`
+
 func main() {
 	if err := run(); err != nil {
 		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
 			fmt.Fprintf(Stdout, "%v\n", err)
 			os.Exit(0)
+		} else if e, ok := err.(*client.Error); ok && e.Message == "access denied" {
```
Comment (niemeyer) [inline cmd/snap/main.go]: Uh oh.. can we have a kind for this?

Comment (niemeyer) [inline cmd/snap/main.go]: Ah, looks like it's a 403, so maybe status code rather than error kind.

Comment (niemeyer) [inline cmd/snap/main.go]: I suggest having a Client.StatusCode and decorating it with what we get in response.StatusCode.

Comment (niemeyer) [inline cmd/snap/main.go]: Actually, 403 or 401 (sorry! :-).

**Comment by mvo5** [comment]
retest this please

**Comment by mvo5** [comment]
retest this please

File: cmd/snap/main.go [hunk 3]
```diff
@@ -130,11 +130,16 @@ func init() {
...
 }
 
+var missingLoginHelp = "Access denied (snap login --help).\n"
+
 func main() {
 	if err := run(); err != nil {
 		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
 			fmt.Fprintf(Stdout, "%v\n", err)
 			os.Exit(0)
+		} else if e, ok := err.(*client.Error); ok && e.StatusCode == 401 && e.Message == "Unauthorized" {
```
Comment (niemeyer) [inline cmd/snap/main.go]: This is still comparing the error message. I suggest making this decision based on whether the status is 401 and we have authorization data at hand (client.LoggedIn?). If we are logged in, pointing out the help won't help much. If we aren't logged in, a 401 is asking us to do so.

File: daemon/daemon.go [hunk 4]
```diff
@@ -116,7 +116,7 @@ func (c *Command) canAccess(r *http.Request) bool {
 
 func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if !c.canAccess(r) {
-		Forbidden("access denied").ServeHTTP(w, r)
+		Unauthorized("Unauthorized").ServeHTTP(w, r)
```
Comment (niemeyer) [inline daemon/daemon.go]: Can we please preserve the "access denied" message here? Although undesired, finding "error: access denied" still feels better than e.g. "error: Unauthorized".

File: integration-tests/tests/auth_test.go [hunk 5]
```diff
@@ -59,11 +59,11 @@ func (s *authSuite) SetUpTest(c *check.C) {
 func (s *authSuite) TestRegressionAuthCrash(c *check.C) {
 	cmd := []string{"snap", "install", "hello-world"}
 	output, _ := cli.ExecCommandErr(cmd...)
-	c.Assert(output, testutil.Contains, `error: access denied`)
+	c.Assert(output, testutil.Contains, `Access denied (snap login --help).`)
```
Comment (niemeyer) [inline integration-tests/tests/auth_test.go]: Comparing the diff, I think this should actually be "error: access denied (snap login --help)", to conform to our other error messages.

**Comment by niemeyer** [comment]
Sorry for the additional trouble on this one.

LGTM with these sorted.

File: daemon/daemon.go (hunk 4 above)
Comment (chipaca) [inline daemon/daemon.go]: We *could* make it `"please login"`

Comment (mvo5) [inline daemon/daemon.go]: I actually like `please login`

File: cmd/snap/main.go (hunk 3 above)
Comment (mvo5) [inline cmd/snap/main.go]: @niemeyer Thanks for this suggestion. One complication here is that the 401 is returned when we try to install something without credentials. But also when we try to log into the store with an invalid password. In both cases we get a 401 and have no credentials stored in the client (yet). 

I pushed 674f3ee01c6d4fca74d30e2c40949cf91df5cb84 that adds an explicit error kind "login-required" that makes it explicit. Please let me know if that looks sensible or if you prefer a different approach to tackle this (we could also add an explicit InvalidCredentials error kind).

File: integration-tests/tests/auth_test.go (hunk 5 above)
Comment (mvo5) [inline integration-tests/tests/auth_test.go]: Thanks, I updated the error string in the client to follow this convention.

File: cmd/snap/cmd_login.go [hunk 6]
```diff
@@ -34,13 +34,21 @@ type cmdLogin struct {
...
 
-var longLoginHelp = i18n.G("This command logs the given username into the store")
+var longLoginHelp = i18n.G(`
+The login command authenticates on snapd and the snap store and saves credentials
+into the ~/.snap/auth.json file. Further communication with snapd will then be made
+using those credentials.
+
+Login only works for local users in the sudo or admin groups.
+
+Credentials may be obtained at https://login.ubuntu.com
```
Comment (pedronis) [inline cmd/snap/cmd_login.go]: An account can be setup?

**Comment by mvo5** [comment]
retest this please

**Comment by niemeyer** [comment]
retest this please

**Comment by niemeyer** [comment]
Unfortunately integration tests are still broken, so it's unclear if this would introduce any regressions or not.

I've picked up that work on #1137 and merged with #1136 to avoid the conflict. I've also changed the approach slightly so it's less dependent on the server. There it will just look at the status code being 401 or 403, and login data being available or not, to decide whether to display the message. The error kind is gone.

Please let me know how you feel.

**Comment by niemeyer** [comment]
After brainstorming with @pedronis, the error kind is coming back.

**Pull request closed without merging**